sarif/issues_report.py

```diff
@@ -0,0 +1,146 @@
+                            zip(common_desc_stem, desc)
+                        ):
+                            if char1 != char2:
+                                key_and_count["key"] = (
+                                    common_desc_stem[0:char_pos] + " …"
+                                )
+                                break
+            sorted_codes = sorted(
+                code_to_key_and_count.keys(),
+                key=lambda code: -code_to_key_and_count[code]["count"],
+            )
+            self._sev_to_sorted_keys[severity] = {
+                code_to_key_and_count[code]["key"]: [] for code in sorted_codes
+            }
+            for record in issues:
+                # Not sorting the issues by location at this point
+                code = record["Code"]
+                self._sev_to_sorted_keys[severity][
+                    code_to_key_and_count[code]["key"]
+                ].append(record)
+
+    def _sort_record_lists(self):
+        if self._sev_to_sorted_keys is None:
+            self._group_records_by_key()
+        for key_to_records in self._sev_to_sorted_keys.values():
+            for records in key_to_records.values():
+                records.sort(key=record_sort_key)
+        self._records_have_been_sorted = True
+
+    def issue_count_for_severity(self, severity: str) -> int:
+        """Get the number of individual records at this severity level."""
+        return len(self._sev_to_records.get(severity, []))
+
+    def issue_type_count_for_severity(self, severity: str) -> int:
+        """Get the number of distinct issue types at this severity level."""
+        if self._sev_to_sorted_keys is None:
+            self._group_records_by_key()
+        return len(self._sev_to_sorted_keys.get(severity, []))
+
+    def any_none_severities(self) -> bool:
+        """Are there any records with severity level "none"?"""
+        return bool(self._sev_to_records.get("none", {}))
+
+    def get_severities(self) -> List[str]:
+        """
+        Get the list of relevant severity levels for these records.
+
+        The returned list always includes "error", "warning" and "note", but "none" is only
+        included at the end if there are any records with severity "none".
+        """
+        return (
+            SARIF_SEVERITIES_WITH_NONE
+            if self.any_none_severities()
+            else SARIF_SEVERITIES_WITHOUT_NONE
+        )
+
+    def get_issues_grouped_by_type(self, severity: str) -> Dict[str, List[dict]]:
+        """
+        Get a dict from issue type key to list of matching records at this severity level.
+
+        Issue type keys are derived from the issue code and (common prefix of) description.
+        """
+        if not self._records_have_been_sorted:
+            self._sort_record_lists()
+        return self._sev_to_sorted_keys.get(severity, {})
+
+    def get_issue_type_histogram(self, severity: str) -> Dict[str, List[dict]]:
```

Comment:
The name `get_issue_type_histogram` seems out of place compared with the others, especially the similar `issue_type_count_for_severity`. Maybe it should be renamed to `get_issue_type_counts_grouped_by_type`? I realize that's a bit long, but it shares structure/terminology with both `get_issues_grouped_by_type` and `issue_type_count_for_severity`.

Reply:
I think histogram is a more expressive way to express it, otherwise it's confusing whether we're counting the issues or the types.  Renamed now to `get_issue_type_histogram_for_severity` for improved consistency with other methods.